In [1]:
! mkdir ~/.kaggle

In [2]:
! cp kaggle.json ~/.kaggle/

In [3]:
! chmod 600 ~/.kaggle/kaggle.json

In [4]:
!kaggle datasets download -d fedesoriano/heart-failure-prediction

  0% 0.00/8.56k [00:00<?, ?B/s]
100% 8.56k/8.56k [00:00<00:00, 7.34MB/s]


In [5]:
! unzip heart-failure-prediction.zip

Archive:  heart-failure-prediction.zip
  inflating: heart.csv               


In [7]:
import pandas as pd

df=pd.read_csv('heart.csv')
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [8]:
df.shape

(918, 12)

Checking features that have outliers

In [10]:
df1 = df[df.Cholesterol<=(df.Cholesterol.mean()+3*df.Cholesterol.std())]
df1.shape

(915, 12)

MaxHR and FastingBS doesn't have outliers

In [11]:
df[df.MaxHR>(df.MaxHR.mean()+3*df.MaxHR.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [12]:
df[df.FastingBS>(df.FastingBS.mean()+3*df.FastingBS.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [13]:
df[df.Oldpeak>(df.Oldpeak.mean()+3*df.Oldpeak.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
166   50   M           ASY        140          231          0         ST   
702   59   M            TA        178          270          0        LVH   
771   55   M           ASY        140          217          0     Normal   
791   51   M           ASY        140          298          0     Normal   
850   62   F           ASY        160          164          0        LVH   
900   58   M           ASY        114          318          0         ST   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
166    140              Y      5.0     Flat             1  
702    145              N      4.2     Down             0  
771    111              Y      5.6     Down             1  
791    122              Y      4.2     Flat             1  
850    145              N      6.2     Down             1  
900    140              N      4.4     Down             1

In [14]:
df2 = df1[df1.Oldpeak<=(df1.Oldpeak.mean()+3*df1.Oldpeak.std())]
df2.shape

(909, 12)

In [15]:
df[df.RestingBP>(df.RestingBP.mean()+3*df.RestingBP.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
109   39   M           ATA        190          241          0     Normal   
241   54   M           ASY        200          198          0     Normal   
365   64   F           ASY        200            0          0     Normal   
399   61   M           NAP        200            0          1         ST   
592   61   M           ASY        190          287          1        LVH   
732   56   F           ASY        200          288          1        LVH   
759   54   M           ATA        192          283          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
109    106              N      0.0       Up             0  
241    142              Y      2.0     Flat             1  
365    140              Y      1.0     Flat             1  
399     70              N      0.0     Flat             1  
592    150              Y      2.0     Down             1  
732    133              Y      4.0     Down             1  
759    195              N      0.0       Up             1

In [16]:
df3 = df2[df2.RestingBP<=(df2.RestingBP.mean()+3*df2.RestingBP.std())]
df3.shape

(902, 12)

In [53]:
df.ChestPainType.unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [54]:
df.RestingECG.unique()

array(['Normal', 'ST', 'LVH'], dtype=object)

In [55]:
df.ExerciseAngina.unique()

array(['N', 'Y'], dtype=object)

In [56]:
df.ST_Slope.unique()

array(['Up', 'Flat', 'Down'], dtype=object)

In [17]:
df4 = df3.copy()
df4.ExerciseAngina.replace(
    {
        'N': 0,
        'Y': 1
    },
    inplace=True)

df4.ST_Slope.replace(
    {
        'Down': 1,
        'Flat': 2,
        'Up': 3
    },
    inplace=True
)

df4.RestingECG.replace(
    {
        'Normal': 1,
        'ST': 2,
        'LVH': 3
    },
    inplace=True)

df4.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40   M           ATA        140          289          0           1   
1   49   F           NAP        160          180          0           1   
2   37   M           ATA        130          283          0           2   
3   48   F           ASY        138          214          0           1   
4   54   M           NAP        150          195          0           1   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         3             0  
1    156               0      1.0         2             1  
2     98               0      0.0         3             0  
3    108               1      1.5         2             1  
4    122               0      0.0         3             0

In [18]:
df5 = pd.get_dummies(df4, drop_first=True)
df5.head()

Age  RestingBP  Cholesterol  FastingBS  RestingECG  MaxHR  ExerciseAngina  \
0   40        140          289          0           1    172               0   
1   49        160          180          0           1    156               0   
2   37        130          283          0           2     98               0   
3   48        138          214          0           1    108               1   
4   54        150          195          0           1    122               0   

   Oldpeak  ST_Slope  HeartDisease  Sex_M  ChestPainType_ATA  \
0      0.0         3             0      1                  1   
1      1.0         2             1      0                  0   
2      0.0         3             0      1                  1   
3      1.5         2             1      0                  0   
4      0.0         3             0      1                  0   

   ChestPainType_NAP  ChestPainType_TA  
0                  0                 0  
1                  1                 0  
2                  0                 0  
3                  0                 0  
4                  1                 0

In [19]:
X=df5.drop('HeartDisease',axis='columns').values
y=df5.HeartDisease

In [20]:
from sklearn.preprocessing import StandardScaler

scaler=StandardScaler()
X_scaled=scaler.fit_transform(X)
X_scaled

array([[-1.42896269,  0.46089071,  0.85238015, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-0.47545956,  1.5925728 , -0.16132855, ..., -0.4836591 ,
         1.86750159, -0.22914788],
       [-1.74679706, -0.10495034,  0.79657967, ...,  2.06757196,
        -0.53547478, -0.22914788],
       ...,
       [ 0.37209878, -0.10495034, -0.61703246, ..., -0.4836591 ,
        -0.53547478, -0.22914788],
       [ 0.37209878, -0.10495034,  0.35947592, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-1.64085227,  0.3477225 , -0.20782894, ..., -0.4836591 ,
         1.86750159, -0.22914788]])

In [21]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X_scaled,y,stratify=y,random_state=10)

In [22]:
X_train.shape

(676, 13)

In [23]:
X_test.shape

(226, 13)

In [24]:
Y_train.value_counts()

1    371
0    305
Name: HeartDisease, dtype: int64

In [25]:
305/371

0.8221024258760108

In [26]:
Y_test.value_counts()

1    124
0    102
Name: HeartDisease, dtype: int64

In [27]:
102/124

0.8225806451612904

Train using stand alone model

In [38]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

clf = GridSearchCV(SVC(gamma='auto'), {
    'C': [1,10,20],
    'kernel': ['rbf','linear']
}, cv=5, return_train_score=False)
clf.fit(X_scaled, y)
clf.cv_results_

{'mean_fit_time': array([0.0205698 , 0.01732011, 0.02220793, 0.06898565, 0.0235815 ,
        0.13416681]),
 'mean_score_time': array([0.00487075, 0.00244012, 0.0044641 , 0.00208597, 0.00447659,
        0.00222368]),
 'mean_test_score': array([0.81590546, 0.80816452, 0.80590546, 0.80705955, 0.80701044,
        0.80705955]),
 'param_C': masked_array(data=[1, 1, 10, 10, 20, 20],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_kernel': masked_array(data=['rbf', 'linear', 'rbf', 'linear', 'rbf', 'linear'],
              mask=[False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'C': 1, 'kernel': 'rbf'},
  {'C': 1, 'kernel': 'linear'},
  {'C': 10, 'kernel': 'rbf'},
  {'C': 10, 'kernel': 'linear'},
  {'C': 20, 'kernel': 'rbf'},
  {'C': 20, 'kernel': 'linear'}],
 'rank_test_score': array([1, 2, 6, 3, 5, 3], dtype=int32),
 'split0_test_score': array([0.84530387, 0.83

In [39]:
df_ = pd.DataFrame(clf.cv_results_)
df_

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0       0.020570      0.005919         0.004871        0.000671       1   
1       0.017320      0.001060         0.002440        0.000629       1   
2       0.022208      0.002322         0.004464        0.000256      10   
3       0.068986      0.006673         0.002086        0.000066      10   
4       0.023582      0.002383         0.004477        0.000240      20   
5       0.134167      0.019844         0.002224        0.000137      20   

  param_kernel                         params  split0_test_score  \
0          rbf      {'C': 1, 'kernel': 'rbf'}           0.845304   
1       linear   {'C': 1, 'kernel': 'linear'}           0.839779   
2          rbf     {'C': 10, 'kernel': 'rbf'}           0.861878   
3       linear  {'C': 10, 'kernel': 'linear'}           0.839779   
4          rbf     {'C': 20, 'kernel': 'rbf'}           0.861878   
5       linear  {'C': 20, 'kernel': 'linear'}           0.839779   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.839779           0.822222           0.805556           0.766667   
1           0.812155           0.822222           0.827778           0.738889   
2           0.823204           0.805556           0.772222           0.766667   
3           0.806630           0.822222           0.827778           0.738889   
4           0.828729           0.800000           0.755556           0.788889   
5           0.806630           0.822222           0.827778           0.738889   

   mean_test_score  std_test_score  rank_test_score  
0         0.815905        0.028304                1  
1         0.808165        0.035768                2  
2         0.805905        0.034947                6  
3         0.807060        0.035713                3  
4         0.807010        0.036076                5  
5         0.807060        0.035713                3

In [40]:
df_[['param_C','param_kernel','mean_test_score']]

param_C param_kernel  mean_test_score
0       1          rbf         0.815905
1       1       linear         0.808165
2      10          rbf         0.805905
3      10       linear         0.807060
4      20          rbf         0.807010
5      20       linear         0.807060

In [41]:
clf.best_params_

{'C': 1, 'kernel': 'rbf'}

In [42]:
from sklearn.model_selection import cross_val_score

scores=cross_val_score(SVC(C=1,kernel='rbf'),X,y,cv=5)
scores

array([0.62430939, 0.79558011, 0.68888889, 0.72777778, 0.61666667])

In [43]:
scores.mean()

0.6906445672191528

Train using Bagging

In [57]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(
    base_estimator=SVC(C=1,kernel='rbf'), 
    n_estimators=100, 
    max_samples=0.8, 
    oob_score=True,
    random_state=0
)
bag_model.fit(X_train, Y_train)
bag_model.oob_score_

0.856508875739645

In [58]:
bag_model.score(X_test, Y_test)

0.8849557522123894

In [59]:
scores = cross_val_score(bag_model, X, y, cv=5)
scores.mean()

0.6839656230816453

Train using stand alone model Decision Tree

In [46]:
from sklearn.tree import DecisionTreeClassifier

scores = cross_val_score(DecisionTreeClassifier(), X, y, cv=5)
scores

array([0.75138122, 0.77348066, 0.76666667, 0.68333333, 0.61666667])

In [47]:
scores.mean()

0.7183057090239411

Train using Bagging with Decision Tree

In [49]:
bag_model = BaggingClassifier(
    base_estimator=DecisionTreeClassifier(), 
    n_estimators=100, 
    max_samples=0.8, 
    oob_score=True,
    random_state=0
)
bag_model.fit(X_train, Y_train)
bag_model.oob_score_

0.849112426035503

In [51]:
bag_model.score(X_test, Y_test)

0.8584070796460177

In [52]:
scores = cross_val_score(bag_model, X, y, cv=5)
scores.mean()

0.8014855739717618

performance increased from 71.8% to 80.1%

Using bagging for SVM doesn't make much difference in terms of model accuracy.Bagging is effective when we have high variance and instable model such as decision tree.

Train using Random Forest which itself uses bagging underneath

In [60]:
from sklearn.ensemble import RandomForestClassifier

scores = cross_val_score(RandomForestClassifier(), X, y, cv=5)
scores.mean()

0.8214548802946592

Random forest gave even a better performance with 82.1% as score.Underneath it used bagging where it sampled not only data rows but also the columns (or features).